src/Storages/StorageS3.h

```diff
@@ -215,8 +266,22 @@ class StorageS3Source : public SourceWithKeyCondition, WithContext
         explicit operator bool() const { return reader != nullptr; }
         PullingPipelineExecutor * operator->() { return reader.get(); }
         const PullingPipelineExecutor * operator->() const { return reader.get(); }
-        String getPath() const { return fs::path(bucket) / key_with_info->key; }
-        const String & getFile() const { return key_with_info->key; }
+        String getPath() const
+        {
+            return key_with_info->path_in_archive.has_value()
+                ? (bucket + "/" + key_with_info->key + "::" + key_with_info->path_in_archive.value())
+                : (bucket + "/" + key_with_info->key);
+        }
+        const String & getFile() const
+        {
+            return key_with_info->path_in_archive.has_value() ? key_with_info->path_in_archive.value() : key_with_info->key;
+        }
+        String getFileExtended() const
```

Comment:
I'd remove this method at all. It is used only for `addNumRowsToCache`, where we anyway add `bucket` to a result of `getFileExtended`, so it is identical to `getPath` method, which we can use instead.